docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+- We cannot set OOM_SCORE_ADJ of a process until it has launched. This could lead to races. For example, suppose that a burstable pod is using 70% of the system’s memory, and another burstable pod is using 30% of the system’s memory. A best-effort pod attempts to launch on the Kubelet. Initially the best-effort pod is using 2% of memory, and has an OOM_SCORE_ADJ of 0. So its OOM_SCORE is lower than the burstable  pod using 70% of system memory. The burstable pod will be evicted by the best-effort pod. Short-term TODO: Consider implementing a restart policy where best-effort pods are immediately evicted if OOM killed, but burstable pods are given a few retries. Long-term TODO: push support for OOM scores in cgroups to the upstream Linux kernel. This is out of scope.
+- We need to deal with the Kubelet’s restart policy when a process gets killed by the kernel. Suppose that we have 30 best-effort pods wanting to use 1GB of memory, with an “always” restart policy, in a system with 8GB of memory. In the current proposal, the system will keep launching best-effort pods, and the kernel will keep killing best-effort pods. In fact, the OOM killer is biased towards killing pods that are consuming more memory, so best-effort pods that are running would be evicted. In the first iteration, we will encourage users not to use the “always” restart policy with best-effort pods, but won’t handle the restart issue in Kubelet. Extension: In the final implementation, we will not allow best-effort pods to restart, and will limit the number of times burstable pods can restart if OOM killed.
+- The OOM killer might kill processes across different pods. Suppose that 10 best-effort pods are running on the Kubelet, and the system runs out of memory. The OOM killer might kill one process from each pod. Short-term TODO: Docker’s philosophy has been not to care about this issue, so for now we will not focus on it. Long-term TODO: Treat a pod as a unit - so that one entire pods is killed, instead of a few processes across different pods.
+- CPU shares allocation issue. Suppose that a node is running 2 pods: a Burstable pod requesting for 50% of CPU (but without a CPU limit), and a Best-Effort pod. Suppose that both pods try to use as much CPU as possible. After the proposal is implemented, the Burstable pod will get 100% of the CPU, and the Best-Effort pod will get around 0% of the CPU. However, a fairer scheme would give the Burstable pod 75% of the CPU and the Best-Effort pod 25% of the CPU (since resources past the Burstable pod’s request are not guaranteed). Long-term TODO: Think of a way to solve this problem.

```

Comment:
What I've previously proposed to solve this problem is dynamically updating the shares of a cgroup containing all the best-effort pods with the shares remaining after allocating shares for all requests. However, if we did that we'd likely need to employ another isolation mechanism to protect the latency of the Burstable and Guaranteed containers, such as core allocation.